### ISSUE TITLE
Twython's update_status_with_media() is deprecated

### ISSUE THREAD

**Issue by mrichardson23** [raspberrypilearning/tweeting-babbage#15]
With regard to:

``` python
with open('/home/pi/image.jpg', 'rb') as photo:
    twitter.update_status_with_media(status=message, media=photo)
```

I've seen an RCE getting the warning that `update_status_with_media` is deprecated in Twython.  The Twython docs show this is the way to upload media:

``` python
photo = open('/path/to/file/image.jpg', 'rb')
response = twitter.upload_media(media=photo)
twitter.update_status(status='Checkout this cool image!', media_ids=[response['media_id']])
```

**Comment by bennuttall** [comment]
I'm aware of this. I think it depends if you install it from pip or apt. pip gives you the latest version (as documented) and apt gives you a slightly older version, for which you cannot browse the documentation. Unfortunately the author does not push version branches of their docs. /cc @waveform80

I think I originally opted for apt because it's the *better* way of installing software. However, if the pip version is documented (and works properly) then we should switch to it and update accordingly.

Have you tested the new method? I've seen it fail for people at Picademy before.

**Comment by waveform80** [comment]
Yeah, `update_status_with_media` is indeed deprecated, but unfortunately the new method (`upload_media`) isn't available in the current Raspbian package (annoyingly the docs aren't versioned either, which caused issues at a picademy a while back as the new docs didn't work with the default installed version - I should find out what happened with the Twython ticket I opened about that when I've got a spare mo to track it down).

For now, leave the worksheet as it is: it works with both the `pip` and `apt` versions of the package. Sure, it raises a DeprecationWarning with the `pip` version but that's acceptable and it's better than having a worksheet that only works with one version.

That said, this ticket is worth leaving open to remind ourselves that we are relying on deprecated functionality and as soon as the Raspbian package catches up with upstream (presumably when the next major release comes along) the worksheet should be updated.

**Comment by mrichardson23** [comment]
Thanks both so much. Really helpful to know this and understand what's going on. I'll be keeping tabs and will be ready if the question comes up again.

**Comment by lawsie** [comment]
Just checking in on this. Has the package caught up yet? If so I can update the resource...

**Comment by bennuttall** [comment]
It won't be updated until we move to Stretch. Jessie version is [3.1.2](https://packages.debian.org/jessie/python-twython), Stretch is [3.4.0](https://packages.debian.org/stretch/python-twython), which matches stretch.

However, we don't refer to the the apt package here any more. As Dave said above, there is no issue except there's a deprecation warning. Should be ok to leave unless we see any issues in future.

**Comment by lawsie** [comment]
@MarcScott Is this relevant to the convo about twython/tweepy?

**Comment by MarcScott** [comment]
Yep - if we're using tweepy for the new resources then this is no longer an issue

**Comment by andypiper** [comment]
Aha - I was just reviewing this material (and the getting started with Twitter API version) and was about to fork and update for the same reason... it would be far better to use the more recent 2-step media upload and then Tweet process and avoid update_with_media. It looks like this is on your radar so I'll leave this for now. 

BTW Thank You for including such a nice Twitter API-related resource in your learning resources! 👍 
LMK how we can help!

**Comment by bennuttall** [comment]
Thanks @andypiper - this kind of project is really popular at our teacher training course, Picademy. It's a really great way to take them from "the button takes a picture" to doing something in the real world - something they can relate to :+1:

**Comment by martinohanlon** [comment]
I am going to close this, the project is now using tweepy, the `update_with_media` is not depreciated in the latest version of tweepy.

**Comment by andypiper** [comment]
update_with_media is deprecated by the Twitter API. I think the latest tweepy supports the two-stage media/upload and statuses/update, but I can't see that reflected in the public documentation. So long as the code used in the class reflects this usage, I don't see an issue.

**Comment by bennuttall** [comment]
Twython supports the two-stage upload/update method:

```python
with open(photo, 'rb') as f:
    response = twitter.upload_media(media=f)
media_ids = [response['media_id']]
twitter.update_status(status=status, media_ids=media_ids)
```

or for multiple photos:

```python
media_ids = []
for i in images:
    with open(photo, 'rb') as f:
        response = twitter.upload_media(media=f)
    media_ids.append(response['media_id'])
tweet_photos(tweet_text, media_ids)
```

I used this method in [jam-photobooth](https://github.com/RaspberryPiFoundation/jam-photobooth).

The issue with `update_with_media` is that there's a deprecation warning (most people think warnings are fatal errors) - and I guess that this could eventually stop being supported by Twitter. Probably best to update to this method.

**Comment by martinohanlon** [comment]
The `update_with_media` method is depreciated in Twython and you do get the warning.

This project has since been changed to use `Tweepy` instead of `Twython` and the `Tweepy.update_with_media` is depreciated. Perhaps internally it makes 2 separate calls? 

Fyi - I have updated [getting started with twitter api](https://github.com/raspberrypilearning/getting-started-with-the-twitter-api/tree/draft) to no longer use the depreciated method as well.
